src/main/java/org/ice4j/util/AsyncQueueHandler.java

```diff
@@ -0,0 +1,459 @@
+import java.util.concurrent.atomic.*;
+import java.util.logging.Logger; // Disambiguation
+
+/**
+ * Asynchronously reads items from provided {@link #queue} on separate thread
+ * borrowed from {@link #executor} and process items with specified handler.
+ * Thread is not blocked when queue is empty and returned back to
+ * {@link #executor} pool. New or existing thread is borrowed from
+ * {@link #executor} when queue is non empty and {@link #reader} is not
+ * running
+ *
+ * @author Yura Yaroshevich
+ */
+public final class AsyncQueueHandler<T>
+{
+    /**
+     * The {@link java.util.logging.Logger} used by the
+     * {@link AsyncQueueHandler} class and its instances for logging output.
+     */
+    private static final java.util.logging.Logger logger
+        = Logger.getLogger(AsyncQueueHandler.class.getName());
+
+    /**
+     * ScheduledExecutorService to implement delay during throttling in
+     * <tt>AsyncQueueHandler</tt>
+     */
+    private final static ScheduledExecutorService delayTimer
+        = Executors.newSingleThreadScheduledExecutor(
+            new CustomizableThreadFactory(
+                AsyncQueueHandler.class.getName() + "-timer-", true));
+
+    /**
+     * The default <tt>ExecutorService</tt> to run <tt>AsyncQueueHandler</tt>
+     * when there is no executor injected in constructor.
+     */
+    private final static ExecutorService sharedExecutor
+        = Executors.newCachedThreadPool(
+            new CustomizableThreadFactory(
+                AsyncQueueHandler.class.getName() + "-executor-", true));
+
+    /**
+     * Number of nanoseconds to consider queue is empty to exit reading loop in
+     * borrowed thread.
+     */
+    private final static long emptyQueueTimeoutNanoseconds = 50;
+
+    /**
+     * Executor service to run {@link #reader}, which asynchronously
+     * invokes specified {@link #handler} on queued items.
+     */
+    private final ExecutorService executor;
+
+    /**
+     * An {@link BlockingQueue <T>} whose items read on separate thread and
+     * processed by provided {@link #handler}.
+     */
+    private final BlockingQueue<T> queue;
+
+    /**
+     * The {@link Handler<T>} used to handle items read from
+     * {@link #queue} by {@link #reader}.
+     */
+    private final Handler<T> handler;
+
+    /**
+     * An identifier of current reader which is used for debugging purpose.
+     */
+    private final String id;
+
+    /**
+     * A flag which indicates if reading of {@link #queue} should be
+     * cancelled.
+     */
+    private final AtomicBoolean cancelled = new AtomicBoolean();
+
+    /**
+     * Synchronization object of current instance state, in particular
+     * used to resolve races between {@link #handleQueueItemsUntilEmpty()}
+     * and {@link #reader} exit. In particular synchronization object used to
+     * access to field {@link #readerFuture} and {@link #delayedFuture}.
+     */
+    private final Object syncRoot = new Object();
```

Comment:
i find the usage of this hard to follow.  you mention here it protects `readerFuture`, but it appears you synchronize on this across a much larger scope in multiple places.  (btw I looked through this after already removing the throttler in my local checkout, so these comments center around the sync'ing of `readerFuture`)

1) in `cancel()` you synchronize on it before calling `cancel(boolean)`.  `cancel(boolean)` does write to readerFuture, but why not just synchronize on `syncRoot` in `cancel(boolean)` directly?  I see you denote that that method assumes the `syncRoot` has already been locked, but why not just do it there?  it'd make it harder for others to make mistakes. 

2) The use in `handleQueueItemsUntilEmpty` seems fine

3) The block in the Runnable's `run` is very big and protects all sorts of things.  As far as I can tell it's only needed in `onYield` and `cancel` (which would be taken care of by #1 above).  I thought perhaps making the lock tighter here would result in acquiring it more than once (in both `onYield` and `cancel`), but we return after calling `onYield` so I don't think that's the case.  EDIT: actually this is wrong.  in `onYield` we call cancel, so we'll lock twice there, but given this case is likely rare it seems a reasonable tradeoff?  My thinking is to try and minimize the likelihood someone could mess up the synchronization in the future.

Reply:
> 1. in cancel() you synchronize on it before calling cancel(boolean). cancel(boolean) does write to readerFuture, but why not just synchronize on syncRoot in cancel(boolean) directly? I see you denote that that method assumes the syncRoot has already been locked, but why not just do it there? it'd make it harder for others to make mistakes.

I've considered nested `synchronized` blocks as unnecessary, that's why I've initially added private methods which are assumed to be executed under synchronized section. I've checked with benchmark that having nested sections does not hurt performance, so I've added it to simplify reading as you've asked.

> 3. The block in the Runnable's run is very big and protects all sorts of things. As far as I can tell it's only needed in onYield and cancel (which would be taken care of by #1 above). I thought perhaps making the lock tighter here would result in acquiring it more than once (in both onYield and cancel), but we return after calling onYield so I don't think that's the case. EDIT: actually this is wrong. in onYield we call cancel, so we'll lock twice there, but given this case is likely rare it seems a reasonable tradeoff? My thinking is to try and minimize the likelihood someone could mess up the synchronization in the future.

I've removed throttling and moved yield out of synchronized block, but poll and cancel still must be under single synchronized block. The reason why poll & cancel must be under **single** synchronized section is that there is inevitable race between `queue.poll` which detects queue is empty and hence reader is going to exit (`cancel` called), and item being added just between `queue.poll` and `cancel` and hence reader must be rescheduled, but `handleQueueItemsUntilEmpty` count not observe that because `Future` is neither null or done at that moment.

Reply:
> I've removed throttling and moved yield out of synchronized block, but poll and cancel still must be under single synchronized block. The reason why poll & cancel must be under single synchronized section is that there is inevitable race between queue.poll which detects queue is empty and hence reader is going to exit (cancel called), and item being added just between queue.poll and cancel and hence reader must be rescheduled, but handleQueueItemsUntilEmpty could not observe that because Future is neither null or done at that moment.

Good catch